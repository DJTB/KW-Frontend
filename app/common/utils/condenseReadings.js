import { groupBy } from 'lodash';

/**
 * Combines kana for each vocab word under a single entry
 * @param  {Array} [readings=[]] vocabulary readings
 * @return {Array} readings with kana combined
 */
export default function condenseReadings(readings = []) {
  // nest readings into groups by character entry
  const groupedReadings = Object.values(groupBy(readings, 'character'));
  // re-order entry with furistring to front
  const primaryFirst = groupedReadings.map((group) =>
    group.sort((entry) => (entry.furigana != null ? -1 : 1)));

  return combineKana(primaryFirst);
}

function combineKana(list) {
  return list.map((entries) =>
    entries
      .map(ensureKanaArray)
      .reduce((entry, next) => (!entry.kana ? next : spreadKana(entry, next)), {}));
}

function spreadKana(entry, next) {
  return {
    ...entry,
    kana: [...entry.kana, ...next.kana],
  };
}

function ensureKanaArray(obj) {
  return {
    ...obj,
    kana: Array.isArray(obj.kana) ? obj.kana : [obj.kana],
  };
}
