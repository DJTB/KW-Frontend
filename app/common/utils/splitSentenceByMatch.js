import { toKatakana, stripOkurigana } from 'wanakana';
import { inflectVerb, inflectAdjective } from './conjugate';
import stripTilde from './stripTilde';

const makeRegex = (toMatch) => new RegExp(`(.*)(${toMatch})(.*)`);

/**
 * @typedef {Object} splitSentenceByMatchReturn
 * @property {string} head - pre-match sentence text
 * @property {string} match - matched string
 * @property {string} tail - post-match sentence text
 */

/**
 * @typedef {Object} splitSentenceByMatchParams
 * @property {String} splitSentenceByMatchParams.sentence - sentence
 * @property {String} splitSentenceByMatchParams.word - vocab word
 * @property {String} splitSentenceByMatchParams.reading - primary vocab word reading
 * @property {String} splitSentenceByMatchParams.verbType - verb type tag IE: 'v1', 'v5', 'v5u'
 * @property {String} splitSentenceByMatchParams.adjType - adj type tag IE: 'adj-i', 'adj-ix'
 */

/**
 * Splits sentence into head, match, & tail.
 * Attempts to match word, hiragana, katakana; or smart partial matches without okurigana.
 * @param {Object} splitSentenceByMatchParams - vocab object
 * @returns {splitSentenceByMatchReturn} - sentence split with match
 * @example
 * splitSentenceByMatch({ sentence: '彼女は私を避けている', word: '避ける', reading: 'さける'});
 * // { head: '彼女は私を', match: '避', tail: 'けている' }
 */
export default function splitSentenceByMatch({
  sentence = '',
  word = '',
  reading = '',
  verbType = '',
  adjType = '',
} = {}) {
  const [cleanWord, cleanReading] = [stripTilde(word), stripTilde(reading)];

  if (cleanWord.length) {
    const wordResult = findMatch(sentence, createMatchers(cleanWord, '', verbType, adjType));
    if (wordResult.match) {
      return wordResult;
    }
  }

  if (cleanReading.length) {
    const readingResult = findMatch(
      sentence,
      createMatchers(cleanWord, cleanReading, verbType, adjType)
    );
    if (readingResult.match) {
      return readingResult;
    }
  }

  return { head: sentence, match: '', tail: '' };
}

function createMatchers(word = '', reading = '', verbType = '', adjType = '') {
  const base = reading || word;
  const matchers = [base];

  if (verbType) {
    matchers.push(
      ...inflectVerb(base, verbType)
        .map(({ form }) => form)
        .sort((a, b) => b.length - a.length)
    );
  } else if (adjType) {
    matchers.push(
      ...inflectAdjective(base, adjType)
        .map(({ form }) => form)
        .sort((a, b) => b.length - a.length)
    );
    // basic stem matching fallback if no verbType or adjType provided
  } else if (reading) {
    // 'づけ' => 'づ' via 漬け
    matchers.push(stripOkurigana(base, { matchKanji: word }));
    // 'おかね' => 'かね' via お金
    matchers.push(stripOkurigana(base, { matchKanji: word, leading: true }));
    // 'つけ' => 'ツケ'
    matchers.push(toKatakana(base));
  } else {
    matchers.push(stripOkurigana(base)); // '飛び込む' => '飛び込'
    matchers.push(stripOkurigana(base, { leading: true })); // 'お金' => '金'
  }

  return [...new Set(matchers)];
}

function findMatch(sentence = '', matchers = []) {
  let matched;
  while (matchers.length) {
    const matcher = matchers.shift();
    matched = sentence.match(makeRegex(matcher));
    if (matched !== null) {
      const [, head, match, tail] = matched;
      return { head, match, tail };
    }
  }
  return {};
}
