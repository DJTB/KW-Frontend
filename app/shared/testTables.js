import { random } from 'lodash';

/**
 * These all assume that the data has been serialized to preferred format in the front end already!
 * This is not the API response shape.
 */

export const levels = [
  {
    level: 1,
    unlocked: false,
    count: 43,
  },
  {
    level: 2,
    unlocked: true,
    count: 90,
  },
  {
    level: 3,
    unlocked: true,
    count: 69,
  },
  {
    level: 4,
    unlocked: false,
    count: 105,
  },
  {
    level: 5,
    unlocked: true,
    count: 126,
  },
  {
    level: 6,
    unlocked: true,
    count: 118,
  },
  {
    level: 7,
    unlocked: true,
    count: 101,
  },
  {
    level: 8,
    unlocked: false,
    count: 136,
  },
  {
    level: 9,
    unlocked: false,
    count: 119,
  },
  {
    level: 10,
    unlocked: false,
    count: 118,
  },
  {
    level: 11,
    unlocked: false,
    count: 126,
  },
  {
    level: 12,
    unlocked: false,
    count: 126,
  },
  {
    level: 13,
    unlocked: false,
    count: 115,
  },
  {
    level: 14,
    unlocked: false,
    count: 116,
  },
  {
    level: 15,
    unlocked: false,
    count: 100,
  },
  {
    level: 16,
    unlocked: false,
    count: 117,
  },
  {
    level: 17,
    unlocked: false,
    count: 124,
  },
  {
    level: 18,
    unlocked: false,
    count: 135,
  },
  {
    level: 19,
    unlocked: false,
    count: 106,
  },
  {
    level: 20,
    unlocked: false,
    count: 114,
  },
  {
    level: 21,
    unlocked: false,
    count: 106,
  },
  {
    level: 22,
    unlocked: false,
    count: 115,
  },
  {
    level: 23,
    unlocked: false,
    count: 97,
  },
  {
    level: 24,
    unlocked: false,
    count: 120,
  },
  {
    level: 25,
    unlocked: false,
    count: 100,
  },
  {
    level: 26,
    unlocked: false,
    count: 124,
  },
  {
    level: 27,
    unlocked: false,
    count: 106,
  },
  {
    level: 28,
    unlocked: false,
    count: 116,
  },
  {
    level: 29,
    unlocked: false,
    count: 107,
  },
  {
    level: 30,
    unlocked: false,
    count: 103,
  },
  {
    level: 31,
    unlocked: false,
    count: 113,
  },
  {
    level: 32,
    unlocked: false,
    count: 111,
  },
  {
    level: 33,
    unlocked: false,
    count: 103,
  },
  {
    level: 34,
    unlocked: false,
    count: 124,
  },
  {
    level: 35,
    unlocked: false,
    count: 100,
  },
  {
    level: 36,
    unlocked: false,
    count: 98,
  },
  {
    level: 37,
    unlocked: false,
    count: 112,
  },
  {
    level: 38,
    unlocked: false,
    count: 100,
  },
  {
    level: 39,
    unlocked: false,
    count: 101,
  },
  {
    level: 40,
    unlocked: false,
    count: 116,
  },
  {
    level: 41,
    unlocked: false,
    count: 107,
  },
  {
    level: 42,
    unlocked: false,
    count: 96,
  },
  {
    level: 43,
    unlocked: false,
    count: 108,
  },
  {
    level: 44,
    unlocked: false,
    count: 103,
  },
  {
    level: 45,
    unlocked: false,
    count: 96,
  },
  {
    level: 46,
    unlocked: false,
    count: 103,
  },
  {
    level: 47,
    unlocked: false,
    count: 94,
  },
  {
    level: 48,
    unlocked: false,
    count: 104,
  },
  {
    level: 49,
    unlocked: false,
    count: 104,
  },
  {
    level: 50,
    unlocked: false,
    count: 87,
  },
  {
    level: 51,
    unlocked: false,
    count: 75,
  },
  {
    level: 52,
    unlocked: false,
    count: 99,
  },
  {
    level: 53,
    unlocked: false,
    count: 103,
  },
  {
    level: 54,
    unlocked: false,
    count: 112,
  },
  {
    level: 55,
    unlocked: false,
    count: 76,
  },
  {
    level: 56,
    unlocked: false,
    count: 88,
  },
  {
    level: 57,
    unlocked: false,
    count: 93,
  },
  {
    level: 58,
    unlocked: false,
    count: 81,
  },
  {
    level: 59,
    unlocked: false,
    count: 79,
  },
  {
    level: 60,
    unlocked: false,
    count: 65,
  },
];

export const vocabs = [
  {
    id: 553,
    history: {
      correct: random(10),
      incorrect: random(10),
    },
    session: {
      correct: random(10),
      incorrect: random(10),
      streak: random(11),
    },
    meanings: ['before long', 'soon'],
    readings: [
      {
        id: 582,
        character: '近々',
        kana: ['ちかぢか'],
        level: 7,
        tags: ['Adverbial Noun'],
        sentenceEn: 'When I saw them on a date, I asked them, “Do I hear wedding bells?”',
        sentenceJa: '彼らがデートをしているのを見かけて近々結婚するのってきいたんだ',
        jlpt: 'JLPT N2',
        common: true,
      },
      {
        id: 583,
        character: '近々',
        kana: ['きんきん'],
        level: 7,
        tags: ['Adverbial Noun'],
        sentenceEn: 'When I saw them on a date, I asked them, “Do I hear wedding bells?”',
        sentenceJa: '彼らがデートをしているのを見かけて近々結婚するのってきいたんだ',
        jlpt: 'JLPT N2',
        common: true,
      },
      {
        id: 584,
        character: '近々',
        kana: ['ちかじか'],
        level: 7,
        tags: ['Adverbial Noun'],
        sentenceEn: 'When I saw them on a date, I asked them, “Do I hear wedding bells?”',
        sentenceJa: '彼らがデートをしているのを見かけて近々結婚するのってきいたんだ',
        jlpt: 'JLPT N2',
        common: true,
      },
      {
        id: 783,
        character: '軈て',
        kana: ['やがて'],
        level: 61,
        tags: ['Adverb'],
        sentenceEn: 'By and by he forgot that affair',
        sentenceJa: 'やがて彼はその事件のことを忘れた',
        jlpt: 'JLPT N2',
        common: true,
      },
    ],
    synonyms: [
      {
        id: 5582,
        character: '間もなく',
        kana: ['まもなく'],
        level: 22,
        tags: ['Adverb'],
        sentenceEn: 'Coming up, Jiyuugaoka Station.',
        sentenceJa: '間もなく、自由が丘駅',
        jlpt: 'JLPT N2',
        common: true,
      },
    ],
  },
  {
    id: 176,
    history: {
      correct: random(10),
      incorrect: random(10),
    },
    session: {
      correct: random(10),
      incorrect: random(10),
      streak: random(11),
    },
    meanings: ['young lady', 'little girl', 'maiden', 'daughter', 'girl'],
    readings: [
      {
        id: 188,
        character: '少女',
        kana: ['しょうじょ'],
        level: 3,
        tags: ['Noun'],
        sentenceEn: "I am touched by the girl's acute sensitivity.",
        sentenceJa: '少女の鋭い感性に触れている',
        jlpt: 'JLPT N3',
        common: true,
      },
      {
        id: 189,
        character: '少女',
        kana: ['おとめ'],
        level: 3,
        tags: ['Noun'],
        sentenceEn: "I am touched by the girl's acute sensitivity.",
        sentenceJa: '少女の鋭い感性に触れている',
        jlpt: 'JLPT N3',
        common: true,
      },
    ],
    synonyms: [],
  },
  {
    id: 280,
    history: {
      correct: random(10),
      incorrect: random(10),
    },
    session: {
      correct: random(10),
      incorrect: random(10),
      streak: random(11),
    },
    meanings: ['young boy', 'juvenile', 'youth', 'boy'],
    readings: [
      {
        id: 298,
        character: '少年',
        kana: ['しょうねん'],
        level: 4,
        tags: ['Noun'],
        sentenceEn: 'The shy boy murmured his name.',
        sentenceJa: 'その内気な少年はぼそぼそと自分の名前をつぶやいた全ての少年少女はを知っています',
        jlpt: 'JLPT N3',
        common: true,
      },
    ],
    synonyms: [],
  },
  {
    id: 424,
    history: {
      correct: random(10),
      incorrect: random(10),
    },
    session: {
      correct: random(10),
      incorrect: random(10),
      streak: random(11),
    },
    meanings: [
      'whole life',
      'entire life',
      'a lifetime',
      'all through life',
      'one existence',
      'a generation',
      'an age',
    ],
    readings: [
      {
        id: 449,
        character: '一生',
        kana: ['いっしょう'],
        level: 5,
        tags: ['Noun', 'No adjective'],
        sentenceEn: 'She worked hard.',
        sentenceJa: '彼女は一生懸命働いた',
        jlpt: 'JLPT N3',
        common: true,
      },
    ],
    synonyms: [],
  },
  {
    id: 502,
    history: {
      correct: random(10),
      incorrect: random(10),
    },
    session: {
      correct: random(10),
      incorrect: random(10),
      streak: random(11),
    },
    meanings: ['honest'],
    readings: [
      {
        id: 529,
        character: '正直',
        kana: ['しょうじき'],
        level: 6,
        tags: ['Noun', 'Na adjective'],
        sentenceEn: 'Him, honest? What a joke!',
        sentenceJa: '正直が聞いて呆れるよ',
        jlpt: 'JLPT N3',
        common: true,
      },
    ],
    synonyms: [],
  },
];

export const readings = {
  single: [
    {
      id: 351,
      character: '売り上げ',
      kana: ['うりあげ'],
      level: 5,
      tags: ['Noun'],
      sentenceEn: 'Sales are down this month.',
      sentenceJa: '今月は売上が落ちた',
      jlpt: 'JLPT N2',
      common: true,
    },
  ],
  sameCharacters: [
    {
      id: 582,
      character: '近々',
      kana: ['ちかぢか'],
      level: 7,
      tags: ['Noun'],
      sentenceEn: 'When I saw them on a date, I asked them, “Do I hear wedding bells?”',
      sentenceJa: '彼らがデートをしているのを見かけて近々結婚するのってきいたんだ',
      jlpt: 'JLPT N2',
      common: true,
    },
    {
      id: 583,
      character: '近々',
      kana: ['きんきん'],
      level: 7,
      tags: ['Noun'],
      furigana: '0:きんきん',
      sentenceEn: 'When I saw them on a date, I asked them, “Do I hear wedding bells?”',
      sentenceJa: '彼らがデートをしているのを見かけて近々結婚するのってきいたんだ',
      jlpt: 'JLPT N2',
      common: true,
    },
    {
      id: 584,
      character: '近々',
      kana: ['ちかじか'],
      level: 7,
      tags: ['Noun'],
      sentenceEn: 'When I saw them on a date, I asked them, “Do I hear wedding bells?”',
      sentenceJa: '彼らがデートをしているのを見かけて近々結婚するのってきいたんだ',
      jlpt: 'JLPT N2',
      common: true,
    },
  ],
  suffix: [
    {
      id: 2476,
      character: '各〜',
      kana: ['かく'],
      level: 22,
      tags: ['Prefix'],
      sentenceEn: 'First, choose a representative for each team.',
      sentenceJa: 'まず始めに、各チームの代表者を決めて下さい',
      jlpt: null,
      common: null,
    },
  ],
  prefix: [
    {
      id: 6337,
      character: '〜漬け',
      kana: ['づけ'],
      level: 59,
      tags: ['Suffix'],
      sentenceEn: 'These are lightly pickled cucumbers.',
      sentenceJa: '胡瓜の浅漬けならありますよ',
      jlpt: null,
      common: null,
    },
    {
      id: 130,
      character: '〜才',
      kana: ['さい'],
      level: 2,
      tags: ['Suffix'],
      sentenceEn: 'I am five years old.',
      sentenceJa: 'わたしは５才です',
      jlpt: null,
      common: null,
    },
  ],
  mixed: [
    {
      id: 933,
      character: 'ハート形',
      kana: ['ハートがた'],
      level: 9,
      tags: ['Noun'],
      sentenceEn: 'When flamingos kiss, their heads and necks form a heart shape.',
      sentenceJa: 'フラミンゴがキスをする時、二羽の頭と首がハート形を描く',
      jlpt: null,
      common: false,
    },
    {
      id: 30,
      character: 'フランス人',
      kana: ['ふらんすじん'],
      level: 1,
      tags: ['Noun'],
      sentenceEn: 'He is French.',
      sentenceJa: '彼はフランス人です',
      jlpt: null,
      common: false,
    },
    {
      id: 3190,
      character: '迷惑メール',
      kana: ['めいわくメール'],
      level: 29,
      tags: ['Noun'],
      sentenceEn:
        'He sent about 2.2 billion spam emails over a period of around one and a half years.',
      sentenceJa: '約１年半で約億通の迷惑メールを送りました',
      jlpt: null,
      common: false,
    },
    {
      id: 6193,
      character: '烏龍茶',
      kana: ['ウーロンちゃ'],
      level: 57,
      tags: ['Noun'],
      sentenceEn: 'Oolong tea without ice, please.',
      sentenceJa: '烏龍茶を氷抜きでください',
      jlpt: null,
      common: null,
    },
    {
      id: 4875,
      character: '缶ビール',
      kana: ['かんビール'],
      level: 44,
      tags: ['Noun'],
      sentenceEn: 'Oh, this canned beer is way too good for me.',
      sentenceJa: 'ああ、この缶ビールはマジでうますぎる',
      jlpt: null,
      common: false,
    },
  ],
  onlyKana: [
    {
      id: 9000,
      common: true,
      jlpt: 'JLPT N4',
      character: 'そろそろ',
      kana: ['そろそろ'],
      tags: ['Adverb', "Adverb taking the 'to' particle"],
      sentenceJa: '「あの、かりんさん...、そろそろお開きにしませんか？」「駄目、勝ち逃げ禁止！今度こそボクが勝つの！」',
      sentenceEn:
        "“Er, Karin ..., shouldn't we call it a night soon?” “No! No quitting while ahead! Next time I'll win for sure!”",
    },
    {
      id: 9001,
      common: false,
      jlpt: null,
      character: 'マンコ',
      kana: ['マンコ'],
      tags: ['Noun'],
      sentenceJa: 'マンコいいな〜',
      sentenceEn: 'What a great pussy.',
    },
  ],
  partialKanaSentence: [
    {
      id: 8800,
      common: true,
      jlpt: 'JLPT N4',
      character: '付く',
      kana: ['つく'],
      tags: ['Transitive Verb'],
      sentenceJa: 'この「オンボード」というのは平たく言えば「グラフィックカードがついていない」ものだ',
      sentenceEn:
        'Frankly speaking, this “on board” means “doesn’t have a graphics card”.',
    },
  ],
};
