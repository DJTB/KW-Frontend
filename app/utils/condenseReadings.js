import groupBy from 'lodash/groupBy';

/**
 * Combines kana for a given character into a single entry
 * @param  {Array} [readings=[]] vocabulary readings
 * @return {Array} readings with kana combined
 */
export default function condenseReadings(readings = []) {
  const groupedReadings = Object.values(groupBy(readings, 'character'));
  return groupedReadings.map((entries) =>
    entries.reduce((entry, next) => {
      if (!entry.kana) {
        return nestKana(next);
      }
      entry.kana.push(next.kana);
      return entry;
    }, {}),
  );
}

/**
 * Removes kana from entry, and nests as a new entry in kana
 * @param  {Object} reading full vocab reading entry
 * @return {Object} vocab reading entry with kana added to nested kana
 */
function nestKana(reading) {
  const { kana, ...rest } = reading;
  return {
    ...rest,
    kana: [kana],
  };
}
