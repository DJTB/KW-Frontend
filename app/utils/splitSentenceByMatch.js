import stripOkurigana from 'kanawana/stripOkurigana';
import toKatakana from 'kanawana/toKatakana';
import tokenize from 'kanawana/tokenize';

import stripTilde from './stripTilde';

// [x, y] => 'x|y'
const joinPipe = arr => arr.join('|');

/**
 * Splits sentence into head, match, & tail.
 * Attempts to match characters, hiragana, katakana; or smart partial matches without okurigana.
 * @param  {String} [sentence=''] - sentence to attempt match
 * @param  {String} [character=''] - japanese characters to attempt to match
 * @param  {String} [kana=''] - japanese kana to attempt to match
 * @return {Object} head, match, and tail
 * @example
 * splitSentenceByMatch('彼女は私を避けている', '避ける', 'さける');
 * // { head: '彼女は私を', match: '避', tail: 'けている' }
 */
export default function splitSentenceByMatch(sentence = '', character = '', kana = ['']) {
  const cleanChars = stripTilde(character); // '〜漬け' => '漬け'
  const strippedChars = stripOkurigana(cleanChars); // '飛び込む' => '飛び込'
  const onlyChars = stripOkurigana(cleanChars, { all: true }); // '売り上げ' => '売上'
  const cleanKana = kana.map(stripTilde); // ['〜つけ'] => ['つけ']
  const cleanKata = cleanKana.map(toKatakana); // ['〜つけ'] => ['ツケ']
  const stripKana = k => k.replace(tokenize(cleanChars).pop(), '') || k;
  const strippedKana = cleanKana.map(stripKana); // ['〜つけ'] => ['つ']

  // '売り上げ' => /(.*?)(売り上げ|売り上|売上|うりあげ|ウリアゲ|うりあ)(.*)/
  const regex = new RegExp(`(.*?)(${
    joinPipe([cleanChars, strippedChars, onlyChars])
  }|${
    joinPipe([cleanKana, cleanKata, strippedKana].map(joinPipe))
  })(.*)`);

  const failSafe = ['', sentence, '', ''];

  const [, head, match, tail] = sentence.match(regex) || failSafe;

  return {
    head,
    match,
    tail,
  };
}
