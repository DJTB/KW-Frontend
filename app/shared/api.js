import urljoin from 'url-join';
import { get, put, post, patch, del } from 'utils/request';

const BASE_URL = '//localhost:8000'; // FIXME: change for production
const API_BASE = urljoin(BASE_URL, 'api', 'v1');

//-----------------------------------------------------------------------------
//  AUTHORIZATION
//-----------------------------------------------------------------------------
const authUrl = urljoin(API_BASE, 'auth');
const userLoginUrl = urljoin(authUrl, 'login');
const userCredentialsUrl = urljoin(authUrl, 'me'); // GET / PATCH
const registerUrl = urljoin(authUrl, 'register'); // POST
const activateUrl = urljoin(authUrl, 'activate'); // POST
const usernameUrl = urljoin(authUrl, 'username'); // POST
const passwordUrl = urljoin(authUrl, 'password'); // POST
const resetPasswordUrl = urljoin(passwordUrl, 'reset'); // POST
const confirmPasswordUrl = urljoin(resetPasswordUrl, 'confirm'); // POST

//-----------------------------------------------------------------------------
//  USER
//-----------------------------------------------------------------------------
const userUrl = urljoin(API_BASE, 'user'); // GET all users (if admin, else 'me')
const userProfileUrl = urljoin(userUrl, 'me'); // GET user profile
const userSrsUrl = urljoin(userUrl, 'srs'); // POST to get review count
const userSyncUrl = urljoin(userUrl, 'sync'); // POST to sync with WK

//-----------------------------------------------------------------------------
//  REVIEWS
//-----------------------------------------------------------------------------
const reviewsUrl = urljoin(API_BASE, 'review'); // GET all ready reviews
const criticalReviewsUrl = urljoin(reviewsUrl, 'critical'); // GET critical
const currentReviewsUrl = urljoin(reviewsUrl, 'current'); // GET current review queue
const currentLessonsUrl = urljoin(reviewsUrl, 'lesson'); // GET current lesson queue
const reviewEntryUrl = (id) => urljoin(reviewsUrl, id); // GET single
const reviewCorrectUrl = (id) => urljoin(reviewEntryUrl(id), 'correct'); // POST correct answer
const reviewIncorrectUrl = (id) => urljoin(reviewEntryUrl(id), 'incorrect'); // POST incorrect answer
const lockReviewUrl = (id) => urljoin(reviewEntryUrl(id), 'hide'); // POST
const unlockReviewUrl = (id) => urljoin(reviewEntryUrl(id), 'unhide'); // POST

//-----------------------------------------------------------------------------
//  SYNONYMS
//-----------------------------------------------------------------------------
const synonymUrl = urljoin(API_BASE, 'synonym'); // POST add, GET get all
const synonymEntryUrl = (id) => urljoin(synonymUrl, id); // DELETE remove, GET one

//-----------------------------------------------------------------------------
//  VOCABULARY
//-----------------------------------------------------------------------------
const vocabularyUrl = urljoin(API_BASE, 'vocabulary'); // GET everything!
const vocabularyEntryUrl = (id) => urljoin(vocabularyUrl, id); // GET one
const readingUrl = urljoin(API_BASE, 'reading'); // GET all
const readingEntryUrl = (id) => urljoin(readingUrl, id); // GET one
const levelsUrl = urljoin(API_BASE, 'level'); // GET all
const levelEntryUrl = (id) => urljoin(levelsUrl, id); // GET one
const lockLevelUrl = (id) => urljoin(levelEntryUrl(id), 'lock'); // POST lock
const unlockLevelUrl = (id) => urljoin(levelEntryUrl(id), 'unlock'); // POST unlock

//-----------------------------------------------------------------------------
//  GENERAL
//-----------------------------------------------------------------------------
const faqUrl = urljoin(API_BASE, 'faq'); // GET all, POST create
const faqEntryUrl = (id) => urljoin(faqUrl, id); // GET, PUT update, PATCH partial update, DELETE
const announcementUrl = urljoin(API_BASE, 'announcement'); // GET all, POST create
const announcementEntryUrl = (id) => urljoin(announcementUrl, id); // GET one, PUT update, PATCH  partial update, DELETE
const contactUrl = urljoin(API_BASE, 'contact');

//-----------------------------------------------------------------------------
//  EXTERNAL
//-----------------------------------------------------------------------------
export const createJishoApiUrl = (keyword) => `//jisho.org/api/v1/search/words?keyword=${keyword}`;
export const createJishoUrl = (keyword) => `//jisho.org/search/${keyword}`;


//-----------------------------------------------------------------------------
//  AUTHORIZATION
//-----------------------------------------------------------------------------
export const getUserAuth = ({ id, username, email } = {}) => get(userCredentialsUrl, { id, username, email });
export const loginUser = ({ username, password } = {}) => post(userLoginUrl, { username, password });
export const updateUserAuth = ({ id, username, email } = {}) => patch(userCredentialsUrl, { id, username, email });
export const registerUser = ({ email, username, password, apiKey } = {}) =>
  post(registerUrl, { email, username, password, api_key: apiKey });

export const activateUser = ({ uid } = {}) => post(activateUrl, { uid });
export const changeUsername = ({ username } = {}) => post(usernameUrl, { username });
export const changePassword = ({ password } = {}) => post(passwordUrl, { password });
export const resetPassword = ({ email } = {}) => post(resetPasswordUrl, { email }); // send password reset email.
export const confirmPassword = () => post(confirmPasswordUrl); // finish reset password process

//-----------------------------------------------------------------------------
//  USER
//-----------------------------------------------------------------------------
export const getUsers = () => get(userUrl);
export const getUserProfile = () => get(userProfileUrl);
export const syncKw = () => post(userSrsUrl);
// true to force ALL users to sync with WK
export const syncWk = ({ fullSync = false } = {}) => post(userSyncUrl, { full_sync: fullSync });

//-----------------------------------------------------------------------------
//  REVIEWS
//-----------------------------------------------------------------------------
export const getReviews = ({
  id,
  offset,
  limit,
  meaningContains,
  srsLevel,
  srsLevelLt,
  srsLevelGt,
} = {}) => get(
  reviewsUrl,
  {
    level: id,
    offset,
    limit,
    meaning_contains: meaningContains,
    srs_level: srsLevel,
    srs_level_lt: srsLevelLt,
    srs_level_gt: srsLevelGt,
  }
);

export const getCriticalReviews = ({ offset, limit } = {}) => get(criticalReviewsUrl, { offset, limit });
export const getCurrentReviews = ({ offset, limit } = {}) => get(currentReviewsUrl, { offset, limit });
export const getCurrentLessons = ({ offset, limit } = {}) => get(currentLessonsUrl, { offset, limit });

export const getReviewEntry = ({ id }) => get(reviewEntryUrl(id));
export const recordReview = ({ id, isCorrect, previouslyIncorrect }) => {
  const body = {
    user_specific_id: id,
    user_correct: isCorrect,
    wrong_before: previouslyIncorrect,
  };
  return isCorrect ? post(reviewCorrectUrl(id), body) : post(reviewIncorrectUrl(id), body);
};

export const lockReview = ({ id }) => post(lockReviewUrl(id));
export const unlockReview = ({ id }) => post(unlockReviewUrl(id));
export const saveReviewNotes = ({ id, notes }) => patch(reviewEntryUrl(id), { notes });

//-----------------------------------------------------------------------------
//  SYNONYMS
//-----------------------------------------------------------------------------
export const addSynonym = ({ reviewId, character, kana } = {}) => post(
  synonymUrl,
  { review: reviewId, character, kana },
);
export const removeSynonym = ({ id }) => del(synonymEntryUrl(id));

//-----------------------------------------------------------------------------
//  VOCABULARY
//-----------------------------------------------------------------------------
export const getVocabulary = ({
  id: level,
  offset,
  limit,
  meaningContains,
  readingsKanaContains,
  readingsCharacterContains,
} = {}) => get(
  vocabularyUrl,
  {
    level,
    offset,
    limit,
    meaning_contains: meaningContains,
    readings_kana_contains: readingsKanaContains,
    readings_character_contains: readingsCharacterContains,
  }
);

export const getLevels = () => get(levelsUrl);

export const getVocabularyEntry = ({ id }) => get(vocabularyEntryUrl(id));

export const getReadings = ({ offset, limit } = {}) => get(readingUrl, { offset, limit });
export const getReadingEntry = ({ id }) => get(readingEntryUrl(id));

export const lockLevel = ({ id }) => post(lockLevelUrl(id));
export const unlockLevel = ({ id }) => post(unlockLevelUrl(id));

//-----------------------------------------------------------------------------
//  GENERAL
//-----------------------------------------------------------------------------
export const getFaqs = () => get(faqUrl);
export const getFaq = ({ id }) => get(faqEntryUrl(id));
export const addFaq = ({ question, answer } = {}) => post(faqUrl, { question, answer });
export const updateFaq = ({ question, answer } = {}) => put(faqUrl, { question, answer });

export const getAnnouncements = () => get(announcementUrl);
export const getAnnouncement = ({ id }) => get(announcementEntryUrl(id));
export const addAnnouncement = ({ title, body } = {}) => post(announcementUrl, { title, body });
export const updateAnnouncement = ({ title, body } = {}) => put(announcementUrl, { title, body });

export const sendContactMessage = ({ name, email, body } = {}) => post(contactUrl, { name, email, body });
