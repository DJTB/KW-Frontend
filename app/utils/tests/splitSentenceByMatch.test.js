import { readings } from 'shared/testTables';
import { flatMap } from 'lodash';
import condenseReadings from 'utils/condenseReadings';

import splitSentenceByMatch from '../splitSentenceByMatch';

describe('splitSentenceByMatch()', () => {
  it('should have a safe default', () => {
    expect(splitSentenceByMatch()).toEqual({ head: '', match: '', tail: '' });
  });

  it('should match testTables', () => {
    flatMap(Object.values(readings), condenseReadings)
      .forEach(({ sentenceJa, character, kana }) => {
        expect(splitSentenceByMatch(sentenceJa, character, kana)).toMatchSnapshot();
      });
  });
});
