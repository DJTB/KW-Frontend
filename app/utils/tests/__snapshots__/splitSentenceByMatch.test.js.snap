// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`splitSentenceByMatch() should match testTables 1`] = `
Object {
  "head": "今月は",
  "match": "売上",
  "tail": "が落ちた",
}
`;

exports[`splitSentenceByMatch() should match testTables 2`] = `
Object {
  "head": "彼らがデートをしているのを見かけて",
  "match": "近々",
  "tail": "結婚するのってきいたんだ",
}
`;

exports[`splitSentenceByMatch() should match testTables 3`] = `
Object {
  "head": "まず始めに、",
  "match": "各",
  "tail": "チームの代表者を決めて下さい",
}
`;

exports[`splitSentenceByMatch() should match testTables 4`] = `
Object {
  "head": "胡瓜の浅",
  "match": "漬け",
  "tail": "ならありますよ",
}
`;

exports[`splitSentenceByMatch() should match testTables 5`] = `
Object {
  "head": "わたしは５",
  "match": "才",
  "tail": "です",
}
`;

exports[`splitSentenceByMatch() should match testTables 6`] = `
Object {
  "head": "フラミンゴがキスをする時、二羽の頭と首が",
  "match": "ハート形",
  "tail": "を描く",
}
`;

exports[`splitSentenceByMatch() should match testTables 7`] = `
Object {
  "head": "彼は",
  "match": "フランス人",
  "tail": "です",
}
`;

exports[`splitSentenceByMatch() should match testTables 8`] = `
Object {
  "head": "約１年半で約億通の",
  "match": "迷惑メール",
  "tail": "を送りました",
}
`;

exports[`splitSentenceByMatch() should match testTables 9`] = `
Object {
  "head": "",
  "match": "烏龍茶",
  "tail": "を氷抜きでください",
}
`;

exports[`splitSentenceByMatch() should match testTables 10`] = `
Object {
  "head": "ああ、この",
  "match": "缶ビール",
  "tail": "はマジでうますぎる",
}
`;

exports[`splitSentenceByMatch() should match testTables 11`] = `
Object {
  "head": "「あの、かりんさん...、",
  "match": "そろそろ",
  "tail": "お開きにしませんか？」「駄目、勝ち逃げ禁止！今度こそボクが勝つの！」",
}
`;

exports[`splitSentenceByMatch() should match testTables 12`] = `
Object {
  "head": "",
  "match": "マンコ",
  "tail": "いいな〜",
}
`;

exports[`splitSentenceByMatch() should match testTables 13`] = `
Object {
  "head": "この「オンボード」というのは平たく言えば「グラフィックカードが",
  "match": "つ",
  "tail": "いていない」ものだ",
}
`;
